src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/UnusedLocalVariableCheck.java

```diff
@@ -0,0 +1,595 @@
+                        && ast.findFirstToken(TokenTypes.LITERAL_THIS) == null) {
+                    checkIdentifierAst(ast.findFirstToken(TokenTypes.IDENT));
+                }
+                break;
+            case TokenTypes.VARIABLE_DEF:
+                if (ast.getParent().getType() != TokenTypes.OBJBLOCK) {
+                    final DetailAST ident = ast.findFirstToken(TokenTypes.IDENT);
+                    final VariableDesc desc = new VariableDesc(ident.getText(), ast);
+                    setScopeOfVariable(ast, desc);
+                    variables.push(desc);
+                }
+                break;
+            case TokenTypes.IDENT:
+                if (!TokenUtil.isOfType(ast.getParent().getType(),
+                        TokenTypes.VARIABLE_DEF, TokenTypes.DOT,
+                        TokenTypes.LITERAL_NEW, TokenTypes.PATTERN_VARIABLE_DEF)) {
+                    checkIdentifierAst(ast);
+                }
+                break;
+            case TokenTypes.LITERAL_INSTANCEOF:
+                final DetailAST patternVariableDef = ast
+                        .findFirstToken(TokenTypes.PATTERN_VARIABLE_DEF);
+                if (patternVariableDef != null) {
+                    final DetailAST identToken = patternVariableDef
+                            .findFirstToken(TokenTypes.IDENT);
+                    final VariableDesc variableDesc =
+                            new VariableDesc(identToken.getText(), patternVariableDef);
+                    setScopeOfPatternVariable(ast, patternVariableDef, variableDesc);
+                    variables.push(variableDesc);
+                }
+                break;
+            default:
+                break;
+        }
+    }
+
+    @Override
+    public void leaveToken(DetailAST ast) {
+        if (TokenUtil.isOfType(ast.getType(),
+                IntStream.concat(Arrays.stream(SCOPES),
+                        Arrays.stream(SCOPES_EXCLUSIVE_TO_PATTERN_VARIABLES)).toArray())) {
+            while (!variables.isEmpty() && variables.peek().getScope() == ast) {
+                final VariableDesc variableDesc = variables.pop();
+                if (!variableDesc.isUsed()) {
+                    final DetailAST currAst = variableDesc.getVariableAst();
+                    log(currAst, MSG_UNUSED_LOCAL_VARIABLE, variableDesc.name);
+                }
+            }
+        }
+    }
+
+    /**
+     * Gets the first DetailAst of type {@link TokenTypes#DOT} or
+     * {@link TokenTypes#METHOD_CALL} token which is the
+     * child of {@link TokenTypes#EXPR}.
+     *
+     * @param exprAst DetailAst of type {@link TokenTypes#EXPR}
+     * @return DetailAst of type {@link TokenTypes#DOT} or {@link TokenTypes#METHOD_CALL}
+     *         if present, otherwise null
+     */
+    private static DetailAST getFirstDotOrMethodCallToken(DetailAST exprAst) {
+        final DetailAST result;
+        final DetailAST dotToken = exprAst.findFirstToken(TokenTypes.DOT);
```

Comment:
Again, `findFirstTokenByPredicate` will be better here.
It looks like we need something like `TokenUtils.findFirstOfAny(DetailAst, int...)`;

Reply:
Did not create a new method, did it with `findFirstTokenByPredicate`